### PR TITLE
Edit Page

### LINKED ISSUE #15: Add ability to edit test page
https://github.com/maxbeatty/jsperf.com/blob/breakdown/_tpl/edit.tpl

### PR DESCRIPTION
**Pull request by asilluron**
Closes #15

### CODE REVIEW

**Comment by asilluron** [comment]
@jsperf/owners Got this to 90% but need a bit of help to work out the JS errors on the page.

**Comment by madskonradsen** [comment]
Maybe I'm mistaken, but doesn't this also close #9 ? :)

**Comment by mathiasbynens** [comment]
I’ll ping @jdalton for the `ui is not defined` on test cases issue.

**Comment by jdalton** [comment]
Looks like a race condition. If I reload the page it may or may not error. I'm not familiar with the current setup is all the non-benchmark code bundled together or loaded separately?

**Comment by jdalton** [comment]
The google url `https://www.google.com/jsapi?autoload=%7Bmodules%3A%5B%7Bname%3A%22visualization%22%2Cversion%3A1%2Cpackages%3A%5B%22corechart%22%2C%22table%22%5D%2Ccallback%3Aui.browserscope.load%7D%5D%7D` is calling `ui.browserscope.load` before `ui` is exported I'm guessing.

**Comment by jdalton** [comment]
An unminified version would help debug.

**Comment by mathiasbynens** [comment]
Published `@mathias/benchmark.js-wrapper-jsperf@1.0.1` unminified.

File: server/repositories/tests.js [hunk 1]
```diff
@@ -17,8 +19,34 @@ module.exports = {
...
+        // Update test
+        if (test.id) {
+          queries.push(db.genericQuery(`UPDATE tests SET title = ${db.escape(test.title)}, defer =  ${db.escape(test.defer)} , code =  ${db.escape(test.code)} WHERE pageID = ${pageID} AND testID = ${test.id}`));
+        } else {
+          queries.push(db.genericQuery(`INSERT INTO ?? (??) VALUES (${pageID}, ${db.escape(test.title)}, ${db.escape(test.defer)}, ${db.escape(test.code)})`, [table, columns]));
+        }
+      }
+    });
+
+    return Promise.each(queries, result => {
```
Comment (maxbeatty) [inline server/repositories/tests.js]: another way to do this without `bluebird`

```
return Promise.all(queries).then((results) => {
  results.forEach((result) => {
    if (result.affectedRows !== 1) {
      throw new Error('Not all tests inserted');
    }
  })
})
```

File: server/web/edit/index.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        .then(function (values) {
+          let page = values[0];
+          page.test = values[1];
+          page.revision = values[2];
+          const own = request.session.get('own') || {};
+          const isOwn = own[page.id];
+          const isAdmin = request.session.get('admin');
+
+          reply.view('edit/index', {
+            benchmark: true,
```
Comment (DevSide) [inline server/web/edit/index.js]: The source of your 'ui.benchmark' error comes from here.
`benchmark: false`

File: server/repositories/tests.js (hunk 1 above)
Comment (mathiasbynens) [inline server/repositories/tests.js]: @asilluron What do you think? I’d prefer going with @maxbeatty’s suggestion.

**Comment by mathiasbynens** [comment]
@asilluron Could you rebase please?

**Comment by mathiasbynens** [comment]
LGTM.

**Pull request merged**